zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1101,7 +1104,10 @@ protected void processStateChanged(WatchedEvent event) {
         fireAllEvents();
       }
     } else if (event.getState() == KeeperState.Expired) {
-      reconnectOnExpiring();
+      _isNewSessionEventFired = false;
+      if (isManagingZkConnection()) {
```

Comment:
As we moved the isManagingZkConnection closer to these methods, it might be cleaner if we just move it inside reconnectOnExpiring() and fireNewSessionEvents(). So we have better protection and the code looks cleaner.

Reply:
As @pkuwm pointed out, processStateChanged is private method always called under lock protection, the same as fireAllEvent, and reconnectionOnExpiring.

Reply:
Yeah, but it is a good practice to check closer to where the real business logic exists.

One example could be, if in the future we want to call reconnectOnExpiring() somewhere else, we don't need to add one more check there.

High-level, these 2 methods shall only be called for non-shared clients. But we don't have a way to block them for shared client implementation. So we just put the check inside, then they won't cause any more problems even if we want to do anything to change the zkclient.

Reply:
Can you be more specific as what you want to do with the change? Sketch how you think about the fireNewSessionEvent, fireAllEvent, reconnectOnExpiring look like? 

To me, there are many ways to put the if and else check. Either way works for me.

Reply:
maybe something like this:

```
private void reconnectOnExpiring() {
  if (!isManagingZkConnection()) {
    return;
  }
  ... // the rest is the same
}
```

```
private void fireNewSessionEvents() {
  if (!isManagingZkConnection()) {
    return;
  }
 ...// the rest is the same.

  }
```

Reply:
Yeah, this is what I meant.

Reply:
done.